src/Transports/MassTransit.EventHubIntegration/Checkpoints/BatchCheckpointer.cs

```diff
@@ -0,0 +1,130 @@
+        {
+            async Task ReceiveAsync()
+            {
+                await _channel.Writer.WriteAsync(confirmation).ConfigureAwait(false);
+            }
+
+            return ReceiveAsync();
+        }
+
+        public async Task Close(ProcessingStoppedReason stoppedReason)
+        {
+            _channel.Writer.Complete();
+
+            if (stoppedReason != ProcessingStoppedReason.Shutdown)
+                _cancellationTokenSource.Cancel();
+
+            await _checkpointTask.ConfigureAwait(false);
+
+            _cancellationTokenSource.Dispose();
+        }
+
+        async Task WaitForBatch()
+        {
+            try
+            {
+                while (await _channel.Reader.WaitToReadAsync(_cancellationTokenSource.Token).ConfigureAwait(false))
+                    await ReadBatch().ConfigureAwait(false);
+            }
+            catch (OperationCanceledException)
+            {
+            }
+            catch (ChannelClosedException)
+            {
+            }
+            catch (Exception exception)
+            {
+                LogContext.Error?.Log(exception, "WaitForBatch Faulted");
+            }
+        }
+
+        async Task ReadBatch()
+        {
+            var timeoutToken = new CancellationTokenSource(_settings.CheckpointInterval);
+            var batchToken = CancellationTokenSource.CreateLinkedTokenSource(timeoutToken.Token, _cancellationTokenSource.Token);
+            var batch = new List<IPendingConfirmation>();
+
+            try
+            {
+                try
+                {
+                    for (var i = 0; i < _settings.CheckpointMessageCount; i++)
+                    {
+                        var confirmation = await _channel.Reader.ReadAsync(batchToken.Token).ConfigureAwait(false);
+
+                        await confirmation.Confirmed.ConfigureAwait(false);
+
+                        batch.Add(confirmation);
+
+                        if (await _channel.Reader.WaitToReadAsync(batchToken.Token).ConfigureAwait(false) == false)
+                            break;
+                    }
+                }
+                catch (OperationCanceledException exception) when (exception.CancellationToken == batchToken.Token && batch.Count > 0)
+                {
+                }
+
+                await _executor.Run(() => Checkpoint(batch)).ConfigureAwait(false);
+            }
+            catch (OperationCanceledException exception) when (exception.CancellationToken == batchToken.Token)
+            {
+            }
+            catch (Exception exception)
+            {
+                for (var i = 0; i < batch.Count; i++)
+                    batch[i].Faulted(exception);
+            }
+            finally
+            {
+                batchToken.Dispose();
+                timeoutToken.Dispose();
+            }
+        }
+
+        Task Checkpoint(IReadOnlyList<IPendingConfirmation> batch)
+        {
+            var last = batch[batch.Count - 1];
```

Comment:
how do you know these are in order?

Reply:
except you are talking about List order they have been added from channel, which is sorted. Apart from that it will be waiting for confirmation in order (by waiting to be completed per each). And channel writer is ordered because, it is writing to channel just after receiving message from kafka consumer, which is sorted